[GTK][WPE] Get render process hardware acceleration information for webkit://gpu from the web process

carlosgcampos opened the pull request:
#### 76be15a26c9e4509861fb87864afdbadf04e5743
<pre>
[GTK][WPE] Get render process hardware acceleration information for webkit://gpu from the web process
<a href="https://bugs.webkit.org/show_bug.cgi?id=301423">https://bugs.webkit.org/show_bug.cgi?id=301423</a>

Reviewed by Nikolas Zimmermann.

We try to create a PlatformDisplay in the UI process with the same
parameters as the process to get the hardware acceleration information.
We could get the information from the web process using an IPC message
to make sure the data we get is accurate and to allow getting
information for other data we don&apos;t currently show because it&apos;s not
available in the UI process.

Canonical link: <a href="https://commits.webkit.org/302185@main">https://commits.webkit.org/302185@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fc6337cf668f7193322627436a1c61dce0e12aeb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128294 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/570 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39126 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/135685 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79763 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/506 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/444 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97650 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65554 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131242 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/336 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114933 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78241 "Found 1 new API test failure: WPE/TestWebKitWebContext:/webkit/WebKitWebContext/no-web-process-leak (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/319 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/78974 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108703 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33524 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138141 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/415 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/393 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106187 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/454 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111274 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105989 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/333 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29824 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/52710 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20039 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/466 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/63453 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/365 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/434 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/430 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebKit/Shared/glib/HardwareAccelerationInfo.h:
```diff
@@ -0,0 +1,45 @@
+
+namespace WebKit {
+
+struct HardwareAccelerationInfo {
```
Maybe makes sense to choose a more generic name from the beginning, such as `WebProcessRenderingInfo` ?

carlosgcampos replied:
I used the naming we currently use in webkit://gpu where there's indeed info about other things that is not gpu specific. For the struct name I'm fine with a more generic name, but I would avoid adding any process name, it doesn't really matter and in the future it could be the gpu process.

nikolaszimmermann replied:
Good point! maybe just something related to rendering settings

nikolaszimmermann commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/DrawingAreaCoordinatedGraphics.cpp:
```diff
@@ -811,4 +811,14 @@ void DrawingAreaCoordinatedGraphics::foreachRegionInDamageHistoryForTesting(Func
+{
+    if (m_layerTreeHost)
+        m_layerTreeHost->fillGLInformation(WTFMove(info), WTFMove(completionHandler));
+    else
```
How can we reach this branch?

carlosgcampos replied:
In GTK when accelerated compositing is disabled.

nikolaszimmermann commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp:
```diff
@@ -416,5 +418,31 @@ void ThreadedCompositor::updateFPSCounter()
 
+void ThreadedCompositor::fillGLInformation(HardwareAccelerationInfo&& info, CompletionHandler<void(HardwareAccelerationInfo&&)>&& completionHandler)
+{
+    if (!m_compositingRunLoop) {
```
Can we actually reach this?

carlosgcampos replied:
This is unlikely, maybe impossible, it can happen if this is called after invalidate(). I could add an assert instead.

nikolaszimmermann replied:
Sounds good

nikolaszimmermann reviewed (approved):
Looks great, r=me with a few questions/comments:

carlosgcampos committed:
[GTK][WPE] Get render process hardware acceleration information for webkit://gpu from the web process
https://bugs.webkit.org/show_bug.cgi?id=301423

Reviewed by Nikolas Zimmermann.

We try to create a PlatformDisplay in the UI process with the same
parameters as the process to get the hardware acceleration information.
We could get the information from the web process using an IPC message
to make sure the data we get is accurate and to allow getting
information for other data we don't currently show because it's not
available in the UI process.

Canonical link: https://commits.webkit.org/302185@main
